core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -470,11 +469,13 @@ protected final void disconnectFromNodeChannel(final Channel channel, final Exce
     }
 
     protected Channel nodeChannel(DiscoveryNode node, TransportRequestOptions options) throws ConnectTransportException {
-        NodeChannels nodeChannels = connectedNodes.get(node);
-        if (nodeChannels == null) {
-            throw new NodeNotConnectedException(node, "Node not connected");
+        try (Releasable ignored = connectionLock.acquire(node.getId())) {
```

Comment:
Yes, I am fully aware that this is a concurrent map and this was never my issue. Your last comment has convinced me to restore the original code though. Thanks for your persistence and the explanation. :)